---
title: Assembly of the Genome in a Bottle HG004 sequencing data
author: Shaun Jackman
output:
  html_document:
    keep_md: true
---

```{r install-packages, eval=FALSE, echo=FALSE}
install.packages("dplyr")
install.packages("ggplot2")
install.packages("ggrepel")
install.packages("knitr")
install.packages("magrittr")
install.packages("readr")
install.packages("rmarkdown")
install.packages("tidyr")
```

```{r load-libraries, message=FALSE}
library(dplyr)
library(ggplot2)
library(ggrepel)
library(knitr)
library(magrittr)
library(readr)
library(tidyr)
```

```{r read-data}
assembly_data_orig <- read_tsv("assembly-stats.tsv")
aligned_data_orig <- read_tsv("samtobreak.tsv")
```

```{r tidy-data}
assembly_data <- assembly_data_orig %>%
	mutate(Directory = dirname(name), File = basename(name)) %>%
		separate(File, c("Species", "Stage", "Ext")) %>%
		select(-Ext)

aligned_data <- aligned_data_orig %>%
	mutate(Directory = dirname(File), File = basename(File)) %>%
		separate(File, c("Reference", "Species", "Stage", "Ext")) %>%
		select(-Reference, -Ext)

assemblers <- list(
	"abyss/k144" = "ABySS 1.9.0",
	"abyss/k144/sealer" = "ABySS 1.9.0 + Sealer",
	"abyss2/k144" = "ABySS 2.0.0",
	"abyss2/k144/sealer" = "ABySS 2.0.0 + Sealer",
	"bcalm" = "BCALM",
	"discovardenovo" = "DISCOVARdenovo",
	"discovardenovo/abyss-scaffold" = "DISCOVARdenovo + ABySS-Scaffold",
	"discovardenovo/besst" = "DISCOVARdenovo + BESST",
	"discovardenovo/links" = "DISCOVARdenovo + LINKS",
	"sga" = "SGA",
	"soapdenovo" = "SOAPdenovo",
	"bionano/aggressive-B2-N2" = "ABySS 1.9.0 + BioNano -xy",
	"abyss/k144/bionano" = "ABySS 1.9.0 + BioNano",
	"abyss/k144/sealer/bionano" = "ABySS 1.9.0 + Sealer + BioNano",
	"abyss2/k144/bionano" = "ABySS 2.0.0 + BioNano",
	"abyss2/k144/sealer/bionano" = "ABySS 2.0.0 + Sealer + BioNano",
	"discovardenovo/bionano" = "DISCOVARdenovo + BioNano",
	"discovardenovo/abyss-scaffold/bionano" = "DISCOVARdenovo + ABySS-Scaffold + BioNano",
	"discovardenovo/besst/bionano" = "DISCOVARdenovo + BESST + BioNano",
	"discovardenovo/links/bionano" = "DISCOVARdenovo + LINKS + BioNano"
)

data <- left_join(aligned_data %>% mutate(Stage = "scaffolds"), assembly_data,
		by = c("Directory", "Species", "Stage")) %>%
		mutate(Directory = factor(Directory, unique(Directory)),
			Stage = factor(Stage, unique(Stage)),
		Name = factor(Directory, levels = names(assemblers), labels = assemblers))

contig_data <- left_join(aligned_data, assembly_data,
		by = c("Directory", "Species", "Stage")) %>%
		mutate(Directory = factor(Directory, unique(Directory)),
			Stage = factor(Stage, unique(Stage)),
		Name = factor(Directory, levels = names(assemblers), labels = assemblers))
```

# Plot scaffold NG50, NGA50 and total breakpoints
```{r plot-scaffold-NGA50-breakpoints}
ggplot(data %>% filter(!grepl("Sealer", Name))) +
	aes(label = Name, x = Total_breakpoints) +
	xlab("Total breakpoints") + ylab("Scaffold NG50 and NGA50 (Mbp)") +
	geom_errorbar(aes(ymin = Scaffold_NGA50 / 1e6, ymax = NG50 / 1e6), width = 50) +
	geom_text_repel(aes(y = NG50 / 1e6 - 0.1),
		nudge_x = -1000, nudge_y = 2, segment.color = "#cccccc") +
	scale_colour_brewer(name = NULL, palette = "Dark2") +
	xlim(0, 4000) +
	theme_grey(28) +
	theme(legend.position = c(.35, .85))
```

# Plot contig NG50, NGA50 and total breakpoints
```{r plot-contig-NGA50-breakpoints}
ggplot(contig_data %>% filter(!grepl("ABySS-Scaffold|BESST|LINKS|BioNano", Name))) +
	aes(label = Name, x = Contig_breakpoints) +
	xlab("Contig breakpoints") + ylab("Contig NG50 and NGA50 (kbp)") +
	geom_errorbar(aes(ymin = Contig_NGA50 / 1e3, ymax = NG50 / 1e3), width = 50) +
	geom_text_repel(aes(y = NG50 / 1e3 - 0.1),
		nudge_x = 200, nudge_y = 2, segment.color = "#cccccc") +
	xlim(0, 4000) +
	theme_grey(28)
```

# Table of metrics
```{r tabulate-metrics}
kable(data)
```

# Table of assembly metrics
```{r tabulate-assembly-metrics}
kable(assembly_data)
```

# Table of aligned metrics
```{r tabulate-aligned-metrics}
kable(aligned_data)
```
